components/ds18b20/src/ds18b20.c

```diff
@@ -0,0 +1,143 @@
+ */
+#include <string.h>
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+#include "esp_check.h"
+#include "onewire_bus.h"
+#include "onewire_cmd.h"
+#include "onewire_crc.h"
+#include "ds18b20.h"
+
+static const char *TAG = "ds18b20";
+
+#define DS18B20_CMD_CONVERT_TEMP      0x44
+#define DS18B20_CMD_WRITE_SCRATCHPAD  0x4E
+#define DS18B20_CMD_READ_SCRATCHPAD   0xBE
+
+/**
+ * @brief Structure of DS18B20's scratchpad
+ */
+typedef struct  {
+    uint8_t temp_lsb;      /*!< lsb of temperature */
+    uint8_t temp_msb;      /*!< msb of temperature */
+    uint8_t th_user1;      /*!< th register or user byte 1 */
+    uint8_t tl_user2;      /*!< tl register or user byte 2 */
+    uint8_t configuration; /*!< resolution configuration register */
+    uint8_t _reserved1;
+    uint8_t _reserved2;
+    uint8_t _reserved3;
+    uint8_t crc_value;     /*!< crc value of scratchpad data */
+} ds18b20_scratchpad_t;
+
+typedef struct ds18b20_device_t {
+    onewire_bus_handle_t bus;
+    onewire_device_address_t addr;
+    uint8_t th_user1;
+    uint8_t tl_user2;
+    ds18b20_resolution_t resolution;
+} ds18b20_device_t;
+
+esp_err_t ds18b20_new_device(onewire_device_t *device, const ds18b20_config_t *config, ds18b20_device_handle_t *ret_ds18b20)
+{
+    ds18b20_device_t *ds18b20 = NULL;
+    ESP_RETURN_ON_FALSE(device && config && ret_ds18b20, ESP_ERR_INVALID_ARG, TAG, "invalid argument");
+    // check ROM ID, the family code of DS18B20 is 0x28
+    if ((device->address & 0xFF) != 0x28) {
+        ESP_LOGD(TAG, "%016llX is not a DS18B20 device", device->address);
+        return ESP_ERR_NOT_SUPPORTED;
+    }
+
+    ds18b20 = calloc(1, sizeof(ds18b20_device_t));
+    ESP_RETURN_ON_FALSE(ds18b20, ESP_ERR_NO_MEM, TAG, "no mem for ds18b20");
+    ds18b20->bus = device->bus;
+    ds18b20->addr = device->address;
+    ds18b20->resolution = DS18B20_RESOLUTION_12B; // DS18B20 default resolution is 12 bits
+
+    *ret_ds18b20 = ds18b20;
+    return ESP_OK;
+}
+
+esp_err_t ds18b20_del_device(ds18b20_device_handle_t ds18b20)
+{
+    ESP_RETURN_ON_FALSE(ds18b20, ESP_ERR_INVALID_ARG, TAG, "invalid argument");
+    free(ds18b20);
+    return ESP_OK;
+}
+
+esp_err_t ds18b20_set_resolution(ds18b20_device_handle_t ds18b20, ds18b20_resolution_t resolution)
+{
+    ESP_RETURN_ON_FALSE(ds18b20, ESP_ERR_INVALID_ARG, TAG, "invalid argument");
+    // reset bus and check if the ds18b20 is present
+    ESP_RETURN_ON_ERROR(onewire_bus_reset(ds18b20->bus), TAG, "reset bus error");
+
+    // send command: DS18B20_CMD_WRITE_SCRATCHPAD
+    uint8_t tx_buffer[10] = {0};
+    tx_buffer[0] = ONEWIRE_CMD_MATCH_ROM;
+    memcpy(&tx_buffer[1], &ds18b20->addr, sizeof(ds18b20->addr));
+    tx_buffer[9] = DS18B20_CMD_WRITE_SCRATCHPAD;
+    ESP_RETURN_ON_ERROR(onewire_bus_write_bytes(ds18b20->bus, tx_buffer, 10), TAG, "send DS18B20_CMD_WRITE_SCRATCHPAD failed");
+
+    // write new resolution to scratchpad
+    tx_buffer[0] = ds18b20->th_user1;
+    tx_buffer[1] = ds18b20->tl_user2;
```

Comment:
I don't know the protocol for it, but the variables `ds18b20->th_user1`, `ds18b20->th_user2` was never changed, is it right?

Reply:
yes, this is part of the **not implemented yet** features (the reason why this component is just 0.1.0).